Check `private` access at resolution, not an invocation

Akirathan opened the pull request:
Fixes #13078

### Pull Request Description

Previously, runtime privacy checking [did not respect lexical scope](https://github.com/enso-org/enso/issues/13078#issuecomment-3028519077). This change moves the privacy checks into [UnresolvedSymbol.resolveFor](https://github.com/enso-org/enso/blob/72ccc501f0fb138003b8d5cb3642ee86a711327e/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedSymbol.java#L77), where the project of the `UnresolvedSymbol`'s scope is compared to the project where the target function is defined.

In other words, private access is checked at resolution (which happens at runtime, before a method is invoked), not at invocation.

### Important Notes
Workarounds introduced in https://github.com/enso-org/enso/commit/08b04dd2300c3d3b2a8fc49d16742c7c85d6f7ee reverted in:
- ed3b0148aace00cbfa8dd7caab0110052a8bc8a7
- e84da7c52970c6942418c7eeb436bb15ab9ed8a6

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [X] The documentation has been updated, if necessary.
- [X] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [X] Unit tests have been written where possible.
- [X] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
Revert 08b04dd2300c3d3b2a8fc49d16742c7c85d6f7ee

Commit:
Revert 08b04dd2300c3d3b2a8fc49d16742c7c85d6f7ee

Commit:
Revert d9cc084c2348e28856d748d15f0bef94c146dc39

Commit:
Revert d9cc084c2348e28856d748d15f0bef94c146dc39

Commit:
Add minimal reproducer

Commit:
test calling private method directly or via lambda

Commit:
Copy projectPrivate property from FunctionSchema when mapping call arguments

Commit:
Revert "Revert d9cc084c2348e28856d748d15f0bef94c146dc39"

This reverts commit 142fecfe02733653c5052bfcc025f0bc1af2e796.

See https://github.com/enso-org/enso/issues/13078#issuecomment-3023442905

Commit:
Revert "Revert 08b04dd2300c3d3b2a8fc49d16742c7c85d6f7ee"

This reverts commit 1a8f5bccf76b017405ff964d7f06db8e6dcf613b.

See https://github.com/enso-org/enso/issues/13078#issuecomment-3023442905

Commit:
Private constructor cannot be called via callback.

See https://github.com/enso-org/enso/issues/13078#issuecomment-3023442905

Commit:
fmt

Commit:
Fix the expected evaluation semantics in tests.

See https://github.com/enso-org/enso/issues/13078#issuecomment-3028519077.

Commit:
Update private methods semantics in tests

GregoryTravis commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/privateaccess/PrivateConstructorAccessTest.java:
```diff
@@ -155,4 +158,94 @@ public void cannotPatternMatchOnPrivateConstructorFromDifferentProject() throws
+            cons "Name"
+
+        main =
+            call_method callback . name
```
If `My_Type` is private, but `My_Type.Cons` is not private, should it still be possible to call `.name` here?

Akirathan replied:
Good point. You are right, it should not be possible to call `. name`, fixed in c80840078091fda2bf1fad32c69c2e464bca1238 to call `.to_text` instead.

GregoryTravis commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/privateaccess/PrivateMethodAccessTest.java:
```diff
@@ -45,4 +55,92 @@ public void typeDoesNotExposePrivateMethodsToPolyglot() {
+        from local.Lib import call_method
+
+        callback priv_method =
+            # Here, we attempt to call `local.Lib.Mod.priv_method`
```
Remove comment? The test succeeds as I would expect.

Akirathan replied:
Leftover comment removed in daa591b966f6f4f924070347015d11680d3e43f2

GregoryTravis reviewed (commented):
Looks good. Just one question about being able to access a field of a public constructor of a private type.

Commit:
Shouldn't be possible to access private field

Commit:
Remove leftover comment

Commit:
Fix encapsulation.md docs

JaroslavTulach commented on docs/semantics/encapsulation.md:
```diff
@@ -149,7 +149,7 @@ main =
 
   # This constructor is private, we have to use factory method.
-  opaque = Closed_Type.Constructor field=42
+  opaque = Closed_Type.factory field=42
```
This might have been an example of a failure. At least next lines indicate the code is supposed to throw runtime exception.

Maybe you just want to change the documentation to clearly says that calling `Closed_Type.Constructor field=42` will fail.

Akirathan replied:
Docs updated in ebdb0fd56ec59520f7ea86d05fc29b6e2631bf9b

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/argument/CallArgumentInfo.java:
```diff
@@ -218,13 +218,17 @@ public FunctionSchema getPostApplicationSchema() {
+              .hasPreapplied(argumentUsed)
+              .oversaturatedArguments(oversaturatedArgInfo)
+              .annotations(originalSchema.getAnnotations());
+      if (originalSchema.isProjectPrivate()) {
```
The only code change is this "inheriting isProjectPrivate"...

Should this be a common pattern then you may want to overload `projectPrivate` and just call `projectPrivate(originalSchema.isProjectPrivate())`.

Akirathan replied:
This is an outdated experiment. Reverted in 289684f0c022a2f312bf9688d1f3b32c0e7b74aa

JaroslavTulach commented on test/Base_Tests/src/Semantic/Private_Spec.enso:
```diff
@@ -62,10 +59,10 @@ add_specs spec_builder =
-            Test.expect_panic Private_Access <|
-                Public_Type.in_ctx <|
-                    Public_Type.cons 42
+        group_builder.specify "can call private constructor from a lambda method" <|
```
- I don't understand how can we can this new behavior based on the simple `CallArgumentInfo` change?
- the name of the test is misleading - we are not calling a constructor, but `cons` method from this lambda

Akirathan replied:
> I don't understand how can we can this new behavior based on the simple CallArgumentInfo change?

As stated in https://github.com/enso-org/enso/pull/13392#discussion_r2204524085, changes in `CallArgumentInfo` were just an experiment that is now outdated and reverted.

> the name of the test is misleading - we are not calling a constructor, but cons method from this lambda

Better test name in 3685ee6f6be4785ac177bd605fcc5856eab676ef

JaroslavTulach commented on test/Base_Tests/src/Semantic/Private_Spec.enso:
```diff
@@ -20,17 +20,14 @@ add_specs spec_builder =
             Public_Type.cons . is_nothing . should_be_false
 
-        group_builder.specify "cannot directly call private constructor" <|
-            Test.expect_panic Private_Access <| Public_Type.cons 42
```
- This test was clearly wrong - calling **public static method** `Public_Type.cons` is certainly allowed
- again the name of the test was misleading - `cons` is a method, not a constructor

JaroslavTulach commented on test/Base_Tests/src/Semantic/Private_Spec.enso:
```diff
@@ -20,17 +20,14 @@ add_specs spec_builder =
         group_builder.specify "can get on hold of private constructor" <|
```
- misleading test name - `Public_Type.cons` is not a constructor, but a method

Akirathan replied:
Better test name in 8f1f229b486ec2968aea402f30561f02f7f0566a

JaroslavTulach commented on test/Base_Tests/src/Semantic/Private_Spec.enso:
```diff
@@ -101,6 +98,11 @@ add_specs spec_builder =
             obj = My_Priv_Type.Cons 42
             obj.ext_method . should_equal 42
 
+        group_builder.specify "can call private method obtained by public getter" <|
```
- yes, this looks like a correct test

JaroslavTulach reviewed (commented):
- looks like the names of the test methods are/were incorrect
- the title needs a change 
   - looking at the CI failures, this is not going to be _minor_
   - I suggest one such title
- a change log entry might be needed

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/privateaccess/PrivateMethodAccessTest.java:
```diff
@@ -45,4 +55,90 @@ public void typeDoesNotExposePrivateMethodsToPolyglot() {
+            new SourceModule(
+                QualifiedName.fromString("Mod"),
+                """
+                private priv_method x = x
```
There is no new test for `.name` syntax. Following should probably be also tested:
```
private 

type A
   V v
 
   private value self = self.v
```
and then get `A.V 42` and call `.value` on it from another project

Akirathan replied:
I believe this case is tested in https://github.com/enso-org/enso/blob/8f1f229b486ec2968aea402f30561f02f7f0566a/test/Base_Tests/src/Semantic/Private_Spec.enso#L81-L84

Or is it not?

JaroslavTulach replied:
- I don't think it is the `.name` syntax test
- the test should call `Path_Helpers.normalize_segments self.parts .name`
- as [this comment suggests](https://github.com/enso-org/enso/pull/13047#discussion_r2084103548)
- I don't see any `.name` call

Akirathan replied:
Reverted workarounds https://github.com/enso-org/enso/commit/08b04dd2300c3d3b2a8fc49d16742c7c85d6f7ee at:
- e84da7c52970c6942418c7eeb436bb15ab9ed8a6
- ed3b0148aace00cbfa8dd7caab0110052a8bc8a7

This definitely tests the `.name` syntax.

JaroslavTulach replied:
Seeing [some of the failures](https://github.com/enso-org/enso/actions/runs/16326276936/job/46117776689?pr=13392#step:14:1313) I'd say it is tested sufficiently. But: Where has your attitude to _write unit tests for everything_ gone?

Commit:
Revert "Copy projectPrivate property from FunctionSchema when mapping call arguments"

This reverts commit 78f2b81bd33f604286979f3bfd35282ae5c792ed.

Commit:
Better test name

Commit:
Better test name

Commit:
Update docs

Commit:
Add test for calling private static method that was directly imported

Commit:
No private checking in InvokeFunctionNode

Commit:
Update tests: Private field cannot be accessed

Commit:
Private check is inside UnresolvedSymbol

Commit:
fmt

Commit:
Fix expected cons name in test

Commit:
Update tests: Private field cannot be accessed

JaroslavTulach commented:
- [PrivateCheckDisabledTest](https://github.com/enso-org/enso/actions/runs/16271672825/job/45941682314?pr=13392#step:14:5596) is failing with

```
  INFO ide_ci::program::command: sbt ℹ️ [error] Test org.enso.interpreter.test.privateaccess.PrivateCheckDisabledTest.privateCtorCanBeAccessedWhenPrivateCheckIsDisabled failed: org.graalvm.polyglot.PolyglotException: Private_Access.Error, took 0.042 sec
 INFO ide_ci::program::command: sbt ℹ️ [error]     at <enso>.Main.main(Main.enso:3)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.graalvm.polyglot.Value.execute(Value.java:1047)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.polyglot.Function.execute(Function.scala:19)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.test.utils.ProjectUtils.testProjectRun(ProjectUtils.java:134)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.test.utils.ProjectUtils.testProjectRun(ProjectUtils.java:89)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.interpreter.test.privateaccess.PrivateCheckDisabledTest.privateCtorCanBeAccessedWhenPrivateCheckIsDisabled(PrivateCheckDisabledTest.java:37)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at java.lang.reflect.Method.invoke(Method.java:565)
```

- [Meta-Value Manipulation](https://github.com/enso-org/enso/runs/45943736431#r2s168)

❌ prevent getting value from private constructor
	Expected an error Private_Access but no error occurred, instead got: (Cons 42) (at Base_Tests/src/Semantic/Meta_Spec.enso:109:13-73).
❌ prevent getting value from constructor in private module
	Expected an error Private_Access but no error occurred, instead got: (Cons 42) (at Base_Tests/src/Semantic/Meta_Spec.enso:120:13-73).

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedSymbol.java:
```diff
@@ -72,13 +78,66 @@ public Pair<Function, Type> resolveFor(Node node, Type type) {
     return null;
   }
 
+  private void ensureIsAccessible(Node node, Function function) throws PanicException {
```
Doing all the checks in `UnresolvedSymbol` seems right.

JaroslavTulach commented on test/Base_Tests/src/Semantic/Private_Spec.enso:
```diff
@@ -17,20 +17,17 @@ add_specs spec_builder =
-        # For more info, see https://github.com/enso-org/enso/issues/6729
-        group_builder.specify "can get reference to private constructor, but cannot call it" <|
+        group_builder.specify "can get reference to private constructor, and call it from different project" <|
             cons_fn data = Public_Type.cons data
```
Where exactly one is getting reference to a private constructor? Shouldn't it be
```ruby
cons_fn = Public_Type.cons
c = Meta.meta cons_fn : Meta.Constructor
c.fields . should_equal []
c.declaring_type . should_fail_with Private_Check
c.name . should_fail...
```
? But yeah, maybe that's happening behind the scene anyway.

JaroslavTulach reviewed (approved):
- return back the code that had to be _changed to lambdas_ in #13047
- having the access checks in `UnresolvedSymbol` seems correct
- please add `.name` test

Commit:
Merge branch 'develop' into wip/akirathan/13078-private-generated-getters

Commit:
Add CHANGELOG entry

Commit:
Do not check private access if private check is disabled

Commit:
Private check in UnresolvedSymbol compares project from correct module scope.

And not from the scope of the current node.

Commit:
Add test for private check of UnresolvedSymbol

Commit:
Merge branch 'develop' into wip/akirathan/13078-private-generated-getters

Commit:
Fix tests in Meta_Spec.

Calling private constructor is possible, there is no more private access check at invocation.

Commit:
Fix test in Private_Spec - it is possible to call private constructor from Java if reference was obtained from public method.
